File: 143-ReorderList/143-ReorderList.py
# Last updated: 9/14/2025, 12:54:09 PM
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        slow = fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        fast = slow.next
        slow.next = None
        while fast:
            fast.next, slow, fast = slow, fast, fast.next
        
        l1 = head
        l2 = slow
        while l2.next:
            l1.next, l1 = l2, l1.next
            l2.next, l2 = l1, l2.next

        

File: 143-ReorderList/143-ReorderList.java
// Last updated: 9/14/2025, 12:54:12 PM
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        ListNode cur = head;
        int n = 0;
        while(cur != null){
            cur = cur.next;
            n++;
        }
        ListNode prev = null;
        cur = head;
        ListNode next = null;
        for(int i = 0; i < n; i++){
            if(i >= n/2){
               next = cur.next;
               cur.next = prev;
               prev = cur;
               cur = next; 
            }
            else cur = cur.next;
        }
        cur = head;
        ListNode tail = prev;
        ListNode next2 = null;
        while(tail.next != null){
            next = cur.next;
            cur.next = tail;
            next2 = tail.next;
            tail.next = next;
            tail = next2;
            cur = next;
        }
    }
}